src/theory/bv/theory_bv_rewrite_rules_operator_elimination.h

```diff
@@ -19,20 +19,22 @@
 
 #pragma once
 
+#include "options/bv_options.h"
 #include "theory/bv/theory_bv_rewrite_rules.h"
 #include "theory/bv/theory_bv_utils.h"
 
 namespace CVC4 {
 namespace theory {
 namespace bv {
 
-template<>
-bool RewriteRule<UgtEliminate>::applies(TNode node) {
+template <>
+inline bool RewriteRule<UgtEliminate>::applies(TNode node)
```

Comment:
I appreciate that C++ has become a maze of twisty little passages but it concerns me when the solution to a problem is to add `inline`.  Maybe we should dig a bit deeper into what is going on here.

Reply:
This is an important point, beyond the inclusion or exclusion of the new rewriting rule in the bit-vector rewriter discussed above. The inline were added not because of the new rule, but because of a unit test that imports `src/theory/bv/theory_bv_rewrite_rules_operator_elimination.h`.
The same problem occurs in my preprocessing pass, since in it I also import the same file and am having the same issue.

My understanding is that this problems occurs just now and not before, because this is the first time `src/theory/bv/theory_bv_rewrite_rules_operator_elimination.h` is being included in more than one file. As the post I linked to earlier (and am linking [again](http://learningcppisfun.blogspot.com/2009/01/inline-specializations-and-multiple.html)), this creates two definitions of the same function, that generates errors of the form:
```
In function ... multiple definition of ... first defined here 
```

The post suggests two possible solutions: either add an `inline` or change the architecture of the rewrite rules, making all `applies` and `apply` functions not template-specializations but overloaded functions, using separate `.h` and `.cpp` files.

The main reason for me choosing the first option is that this seems to be what was chosen in all other cases in the bit-vector rewriter.
The rewriting rules are separated into the following files:
```
src/theory/bv/theory_bv_rewrite_rules_constant_evaluation.h 
src/theory/bv/theory_bv_rewrite_rules_simplification.h          
src/theory/bv/theory_bv_rewrite_rules_operator_elimination.h
src/theory/bv/theory_bv_rewrite_rules_normalization.h       
```

In all of them except for `src/theory/bv/theory_bv_rewrite_rules_operator_elimination.h`, all functions are inlined. In addition, all these functions are implemented as specializations of [this](https://github.com/CVC4/CVC4/blob/2a96d7de381aa565a6eacf724848c0e7839c7cf6/src/theory/bv/theory_bv_rewrite_rules.h#L352) class. Notice that in it, the functions are declared with `inline` (see, e.g., [here](https://github.com/CVC4/CVC4/blob/2a96d7de381aa565a6eacf724848c0e7839c7cf6/src/theory/bv/theory_bv_rewrite_rules.h#L411)), suggesting that the specialization should be inlined too.

Perhaps there is also a third option, but I am not aware of one.

Reply:
Having opened this can of worms I am now going to do the cowardly thing and try to shunt this on to someone else's desk.  I know @mpreiner and @aniemetz have Plans for the BV solver.  Perhaps they have already attacked the refactoring of this part?

Reply:
No we haven't yet. But for me it's ok to apply these changes since it's consistent with the other rewrite rule header files.